sql/partitioning/partition_base.cc

```diff
@@ -4046,6 +4049,31 @@ bool Partition_base::commit_inplace_alter_table(
   if (ha_alter_info->alter_info->flags == Alter_info::ALTER_PARTITION)
     DBUG_RETURN(false);
 
+  std::string full_table_path = std::string("./") + altered_table->s->db.str +
+                                std::string("/") +
+                                altered_table->s->table_name.str;
+  if (ha_alter_info->alter_info->flags & Alter_info::ALTER_DROP_PARTITION) {
+    handler **file = m_file;
+    partition_info *part_info = ha_alter_info->modified_part_info;
+    List_iterator_fast<partition_element> part_it(part_info->partitions);
+    partition_element *part_elem;
+    while ((part_elem = part_it++) != nullptr) {
+      partition_state state = part_elem->part_state;
+      switch (state) {
+        case PART_TO_BE_DROPPED:
+          char name[FN_REFLEN];
+          // TODO(luqun): add subpartition support
+          create_partition_name(name, full_table_path.c_str(),
+                                part_elem->partition_name, false);
+          (*file)->ha_external_lock(get_thd(), F_UNLCK);
+          (*file)->ha_delete_table(name, old_table_def);
+          break;
```

Comment:
how do we handle if one partition couldn't be deleted/dropped? I guess ALTER guarantees no concurrent DMLS.

Reply:
Yes, SQL layer will acquire exclusive lock and clean table cache before commit

https://github.com/facebook/mysql-5.6/blob/5dd1adebbcc81a8db2c8c30eb20c63d484ee68de/sql/sql_table.cc#L12648